File: src/test/java/org/hisp/dhis/rules/RuleVariableValueAssert.java
package org.hisp.dhis.rules;

import org.hisp.dhis.rules.models.RuleValueType;

import javax.annotation.CheckForNull;
import javax.annotation.Nonnull;

import static org.junit.Assert.assertEquals;

final class RuleVariableValueAssert
{
    private final RuleVariableValue variableValue;

    private RuleVariableValueAssert( RuleVariableValue variableValue )
    {
        this.variableValue = variableValue;
    }

    @Nonnull
    static RuleVariableValueAssert assertThatVariable( @Nonnull RuleVariableValue variableValue )
    {
        return new RuleVariableValueAssert( variableValue );
    }

    @Nonnull
    RuleVariableValueAssert hasValue( @CheckForNull String value )
    {
        assertEquals( value , variableValue.value() );
        return this;
    }

    @Nonnull
    RuleVariableValueAssert hasCandidates( @Nonnull String... candidates )
    {
        assertEquals( candidates.length , variableValue.candidates().size() );
        for ( int index = 0; index < candidates.length; index++ )
        {
            assertEquals( candidates[index] , variableValue.candidates().get( index ) );
        }
        return this;
    }

    @Nonnull
    RuleVariableValueAssert isTypeOf( @Nonnull RuleValueType valueType )
    {
        assertEquals( valueType , variableValue.type() );
        return this;
    }
}

File: src/test/java/org/hisp/dhis/rules/RuleVariableValueAssert.kt
package org.hisp.dhis.rules

import org.hisp.dhis.rules.models.RuleValueType

import org.assertj.core.api.Java6Assertions.assertThat

class RuleVariableValueAssert (private val variableValue: RuleVariableValue) {

    fun hasValue(value: String?) = apply { assertThat(variableValue.value).isEqualTo(value) }

    fun hasCandidates(vararg candidates: String) = apply {
        assertThat(variableValue.candidates.size).isEqualTo(candidates.size)
        candidates.forEachIndexed { index, candidate ->
            assertThat(variableValue.candidates[index]).isEqualTo(candidate)
        }
    }

    fun isTypeOf(valueType: RuleValueType) = apply { assertThat(variableValue.type).isEqualTo(valueType) }

    companion object {

        @JvmStatic
        fun assertThatVariable(variableValue: RuleVariableValue) = RuleVariableValueAssert(variableValue)

    }
}
